modules/zip/zip_packer.cpp

```diff
@@ -0,0 +1,102 @@
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "core/io/zip_io.h"
+
+#include "zip_packer.h"
+
+Error ZIPPacker::open(String path, ZipAppend append) {
+	if (fa.is_valid()) {
+		close();
+	}
+
+	zlib_filefunc_def io = zipio_create_io(&fa);
+	zf = zipOpen2(path.utf8().get_data(), append, NULL, &io);
+	return zf != NULL ? OK : FAILED;
+}
+
+Error ZIPPacker::close() {
+	ERR_FAIL_COND_V_MSG(fa.is_null(), FAILED, "ZIPPacker cannot be closed because it is not open.");
+
+	return zipClose(zf, NULL) == ZIP_OK ? OK : FAILED;
+}
+
+Error ZIPPacker::start_file(String path) {
+	ERR_FAIL_COND_V_MSG(fa.is_null(), FAILED, "ZIPPacker must be opened before use.");
+
+	zip_fileinfo zipfi;
+
+	OS::DateTime time = OS::get_singleton()->get_datetime();
+
+	zipfi.tmz_date.tm_hour = time.hour;
+	zipfi.tmz_date.tm_mday = time.day;
+	zipfi.tmz_date.tm_min = time.minute;
+	zipfi.tmz_date.tm_mon = time.month - 1;
+	zipfi.tmz_date.tm_sec = time.second;
+	zipfi.tmz_date.tm_year = time.year;
+	zipfi.dosDate = 0;
+	zipfi.external_fa = 0;
+	zipfi.internal_fa = 0;
+
+	int ret = zipOpenNewFileInZip(zf, path.utf8().get_data(), &zipfi, NULL, 0, NULL, 0, NULL, Z_DEFLATED, Z_DEFAULT_COMPRESSION);
```

Comment:
Don't we need to check if zf is not null, and if that's the case to call `zipCloseFileInZip` before doing the new open ?

Reply:
Thanks for the hint. I added an additional null check.